pyomo/core/expr/expr_pyomo5.py

```diff
@@ -0,0 +1,3434 @@
+# =====================================================
+
+class _SizeVisitor(SimpleExpressionVisitor):
+
+    def __init__(self):
+        self.counter = 0
+
+    def visit(self, node):
+        self.counter += 1
+
+    def finalize(self):
+        return self.counter
+
+
+def _sizeof_expression(expr):
+    """
+    Return the number of nodes in the expression tree.
+
+    Args:
+        expr: The root node of an expression tree.
+
+    Returns:
+        A non-negative integer that is the number of
+        interior and leaf nodes in the expression tree.
+    """
+    visitor = _SizeVisitor()
+    return visitor.xbfs(expr)
+
+# =====================================================
+#  evaluate_expression
+# =====================================================
+
+class _EvaluationVisitor(ExpressionValueVisitor):
+
+    def visit(self, node, values):
+        """ Visit nodes that have been expanded """
+        return node._apply_operation(values)
+
+    def visiting_potential_leaf(self, node):
+        """
+        Visiting a potential leaf.
+
+        Return True if the node is not expanded.
+        """
+        if node.__class__ in nonpyomo_leaf_types:
+            return True, node
+
+        if node.is_variable_type():
+            return True, value(node)
+
+        if not node.is_expression_type():
+            return True, value(node)
+
+        return False, None
+
+
+def evaluate_expression(exp, exception=True):
+    """
+    Evaluate the value of the expression.
+
+    Args:
+        expr: The root node of an expression tree.
+        exception (bool): A flag that indicates whether
+            exceptions are raised.  If this flag is
+            :const:`False`, then an exception that
+            occurs while evaluating the expression
+            is caught and the return value is :const:`None`.
+            Default is :const:`True`.
+
+    Returns:
+        A floating point value if the expression evaluates
+        normally, or :const:`None` if an exception occurs
+        and is caught.
+    """
+    try:
+        visitor = _EvaluationVisitor()
+        return visitor.dfs_postorder_stack(exp)
+
+    except TemplateExpressionError:     #pragma: no cover
+        if exception:
+            raise
+        return None
+    except ValueError:
+        if exception:
+            raise
+        return None
+
+
+# =====================================================
+#  identify_components
+# =====================================================
+
+class _ComponentVisitor(SimpleExpressionVisitor):
+
+    def __init__(self, types):
+        self.seen = set()
+        if types.__class__ is set:
+            self.types = types
+        else:
+            self.types = set(types)
+
+    def visit(self, node):
+        if node.__class__ in self.types:
+            if id(node) in self.seen:
+                return
+            self.seen.add(id(node))
+            return node
+
+
+def identify_components(expr, component_types):
```

Comment:
Is this used by anything?  This seems bad to identify components using `__class__` and not ctype.  Either it should use ctype or not be called "identify_components".

Reply:
This is used in connector expansion.  I'm not sure I agree with your argument about the name.   I think it's reasonable to find a component by its class type.